samples/sample_multi_transcode/src/sample_multi_transcode.cpp

```diff
@@ -358,90 +359,112 @@ void Launcher::DoTranscoding()
 {
     mfxStatus sts = MFX_ERR_NONE;
 
-    // get parallel sessions parameters
-    mfxU32 totalSessions = (mfxU32)m_pSessionArray.size();
-    MSDKThread* pthread = 0;
-
-    for (mfxU32 i = 0; i < totalSessions; i++)
+    // Helper structure to manage transcoding threads/sessions
+    struct ThreadHandle
     {
-        pthread = new MSDKThread(sts, TranscodeRoutine, (void *)m_pSessionArray[i]);
-        m_HDLArray.push_back(pthread);
-    }
+        std::future<void>       future;
+        ThreadTranscodeContext* context;
+
+        ThreadHandle(std::future<void>&& future, ThreadTranscodeContext* context)
+        : future(std::move(future)), context(context)
+        {
+        }
+    };
 
-    // Need to determine overlay threads count first
-    size_t nOverlayThreads = 0;
-    for (size_t i = 0; i < m_pSessionArray.size(); i++)
+    // Run transcoding threads
+    std::vector<ThreadHandle> handles;
+    handles.reserve(m_pSessionArray.size());
+    for (auto session : m_pSessionArray)
     {
-        if (m_pSessionArray[i]->pPipeline->IsOverlayUsed())
-            nOverlayThreads++;
+        handles.emplace_back(std::async(std::launch::async,
+                             TranscodeRoutine, (void *)session), session);
     }
 
     // Transcoding threads waiting cycle
-    size_t aliveSessionsCount = totalSessions;
+    size_t aliveSessionsCount = handles.size();
     while (aliveSessionsCount)
     {
         aliveSessionsCount = 0;
-        for (mfxU32 i = 0; i < totalSessions; i++)
+        size_t aliveOverlaySessionsCount = 0;
+
+        for (size_t i = 0; i < handles.size(); i++)
```

Comment:
I think the idea with handles was to have `for (auto& handle: handles)` here. We can't do that for the reason I miss?

Reply:
Session index is used to print the session number in case of error/warnings (l.407-412, l.424-429)